### PR TITLE
Unit testing core util

### LINKED ISSUE #189: Increase code coverage of unit tests for Preview 1
We need to write more unit tests in order to hit code coverage targets for:
- [x] azure-core
- [ ] ~~azure-identity~~
- [x] azure-storage-blob

The code coverage target is 70% for each framework.

### PR DESCRIPTION
**Pull request by vcolin7**
Thanks for taking the time to contribute. Please take a moment to fill out the information below.

Partially addresses #189.

Changes Proposed in this pull request:
- Increased test coverage for package `com.azure.android.core.util`.

This considers changes made in #215.

### CODE REVIEW

**Comment by bsiegel** [comment]
If possible I'd like to find a way to keep mocking frameworks out of our library. Generally unless it's something coupled to the Android OS we should be able to redesign the internals of whatever it is we're testing to avoid them.

**Comment by vcolin7** [comment]
@bsiegel I'll try to spend some time thinking if we can get around using PowerMock or Mockito for the Base64 classes.

**Comment by vcolin7** [comment]
@bsiegel Also, I'm curious as to why we want to avoid mocking libraries in our tests

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java [hunk 1]
```diff
@@ -14,14 +13,20 @@ private Base64Util() {
         // Empty constructor to prevent instantiation of this class.
     }
 
+    private static Base64Wrapper base64Wrapper = new Base64Android();
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java]: This is for testing purposes only right? If not should this be atomic, or locked when modified, as this isn't completely thread-safe right now.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java [hunk 2]
```diff
@@ -83,6 +88,35 @@ public static String encodeToString(byte[] src) {
      * @return The decoded byte array.
      */
     public static byte[] decodeString(String encoded) {
-        return encoded == null ? null : Base64.decode(encoded, Base64.DEFAULT);
+        return encoded == null ? null : base64Wrapper.decode(encoded, Base64.DEFAULT);
+    }
+
+    interface Base64Wrapper {
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java]: Is this planned to be used in the public API? If so, instead of maintaining context in this static helper class should a new overload to each API be added which accepts this interface? That way this class remains thread safe and stateless as expected with a static helper class.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+public class Base64UrlTest {
+    @BeforeClass
+    public static void setUp() {
+        Base64Util.setBase64Wrapper(new Base64Stub());
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java]: Is a wrapper being set as Base64 isn't consistent between Android versions or between devices?

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        assertEquals(Arrays.hashCode(ENCODED_BYTES), base64Url.hashCode());
+    }
+
+    @SuppressWarnings("SimplifiableJUnitAssertion")
+    @Test
+    public void equalTo_identicalObject() {
+        final Base64Url base64Url = new Base64Url((byte[]) null);
+        final Base64Url otherBase64Url = new Base64Url((byte[]) null);
+
+        assertTrue(base64Url.equals(otherBase64Url));
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java]: Any reason that `assertEquals` can't be used here? It would allow for the suppression to be removed.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    @SuppressWarnings("SimplifiableJUnitAssertion")
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java]: Same question

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+        assertFalse(base64Url.equals(otherBase64Url));
+    }
+
+    @SuppressWarnings({"SimplifiableJUnitAssertion", "ConstantConditions"})
+    @Test
+    public void notEqualTo_nullObject() {
+        final Base64Url base64Url = new Base64Url((byte[]) null);
+
+        assertFalse(base64Url.equals(null));
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java]: Does `Base64Url`'s `equals` overload compare against the `String` representation? If so could this be `assertNull(base64Url.toString())`?

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    @SuppressWarnings({"SimplifiableJUnitAssertion", "EqualsBetweenInconvertibleTypes"})
+    @Test
+    public void notEqualTo_objectFromDifferentClass() {
+        final Base64Url base64Url = new Base64Url((byte[]) null);
+
+        assertFalse(base64Url.equals(""));
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java]: Same comment as `notEqualTo_nullObject`.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UtilTest.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Test
+    public void decodeUrl_fromNullByteArray() {
+        assertNull(Base64Util.decodeUrl((String) null));
+    }
+
+    @Test
+    public void decodeUrl_fromNullString() {
+        assertNull(Base64Util.decodeUrl((String) null));
+    }
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UtilTest.java]: I think some tests should be added that use non-null, right now this will get good code coverage but is missing real world scenarios.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/CoreUtilTest.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    public void isNullOrEmpty_returnsFalse_withNonEmptyString() {
+        assertFalse(isNullOrEmpty("Test"));
+    }
+
+    @Test
+    public void replaceWithSingleCharacters_inString() {
+        Map<Character, CharSequence> pairs = new HashMap<>();
+        pairs.put('a', "A");
+        pairs.put(' ', "_");
+        String testString = "Ohana means family.";
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/CoreUtilTest.java]: ❤️ 😄

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/DateTimeRfc1123Test.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+        assertEquals(TEST_DATE, dateTimeRfc1123.toString());
+    }
+
+    @Test
+    public void getHashCode() {
+        DateTimeRfc1123 dateTimeRfc1123 = new DateTimeRfc1123(
+            OffsetDateTime.of(LocalDateTime.parse(TEST_DATE, RFC1123_DATE_TIME_FORMATTER), ZoneOffset.UTC));
+
+        assertEquals(1468106660, dateTimeRfc1123.hashCode());
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/DateTimeRfc1123Test.java]: Unless we have overridden the hash code method I wouldn't eagerly test this as it is dependent on the underlying JVM and has the potential for collisions.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/DateTimeRfc1123Test.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+    @SuppressWarnings("SimplifiableJUnitAssertion")
+    @Test
+    public void equalTo_identicalObject() {
+        final DateTimeRfc1123 dateTimeRfc1123 = new DateTimeRfc1123(
+            OffsetDateTime.of(LocalDateTime.parse(TEST_DATE, RFC1123_DATE_TIME_FORMATTER), ZoneOffset.UTC));
+        final DateTimeRfc1123 otherDateTimeRfc1123 = new DateTimeRfc1123(
+            OffsetDateTime.of(LocalDateTime.parse(TEST_DATE, RFC1123_DATE_TIME_FORMATTER), ZoneOffset.UTC));
+
+        assertTrue(dateTimeRfc1123.equals(otherDateTimeRfc1123));
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/DateTimeRfc1123Test.java]: `assertEquals`?

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+    @Test
+    public void convertToString() {
+        assertEquals("Test", MyExpandableStringEnum.fromString("Test").toString());
+    }
+
+    @Test
+    public void getHashCode() {
+        assertEquals(-2043209345, MyExpandableStringEnum.fromString("Test").hashCode());
+    }
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java]: Same comment as the other `hashCode` test.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+        MyExpandableStringEnum otherMyExpandableStringEnum = MyExpandableStringEnum.fromString("Another");
+        Collection instances = MyExpandableStringEnum.values();
+
+        assertFalse(instances.isEmpty());
+        assertEquals(2, instances.size());
+
+        Iterator iterator = instances.iterator();
+
+        assertSame(myExpandableStringEnum, iterator.next());
+        assertSame(otherMyExpandableStringEnum, iterator.next());
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java]: How strong is the guarantee that the backing `Map` for `ExandableStringEnum` will return a collection where the elements are ordered by how they are added?

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+        assertFalse(myExpandableStringEnum.equals(otherMyExpandableStringEnum));
+    }
+
+    @SuppressWarnings({"SimplifiableJUnitAssertion", "ConstantConditions"})
+    @Test
+    public void notEqualTo_nullObject() {
+        MyExpandableStringEnum myExpandableStringEnum = MyExpandableStringEnum.fromString("Test");
+
+        assertFalse(myExpandableStringEnum.equals(null));
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java]: `assertNotNull`?

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java (hunk 1 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java]: It is only for testing purposes. I probably should add a clarifying comment though.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java (hunk 2 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java]: This is not planned to be a part of the public API. The reason I wrote this interface was so we wouldn't have to mock the static methods in Android's Base64 class. This is due to a couple of reasons:

- Unfortunately, we cannot use Java's version of this class since it only works with a higher Android version than our supported minimum.
- We cannot access the Android framework's classes in unit tests, which are only accessible when running on the actual operating system on an emulator or physical device. Unit tests run on the local JVM.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java (hunk 4 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java]: There is no reason not to use `assertEquals`. I must have missed it. Thanks for pointing it out!

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java (hunk 5 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java]: See answer above.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/DateTimeRfc1123Test.java (hunk 11 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/DateTimeRfc1123Test.java]: Agreed.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/DateTimeRfc1123Test.java (hunk 10 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/DateTimeRfc1123Test.java]: We did override it in our `DateTimeRfc1123` class, however, it just uses the `hashCode` method in `OffsetDateTime`. I think I will remove this test for the time being.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java (hunk 12 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java]: See comment above for the same question.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java (hunk 14 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java]: Agreed.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java (hunk 13 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/ExpandableStringEnumTest.java]: There is no guarantee since this is not a LinkedHashSet, so I think the best course of action would be to cast the collectionreturned by `values()` to a HashSet and see if the values we are looking for are contained there.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/DateTimeRfc1123Test.java (hunk 10 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/DateTimeRfc1123Test.java]: Actually, after digging around on the internet, I think I will take the approach showed [here](https://stackoverflow.com/questions/4449728/how-can-i-do-unit-test-for-hashcode).

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java (hunk 3 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java]: The reason to set this here is that we cannot call Android framework's classes like Base64 from unit tests and we cannot use Java's Base64 class because that's only available at a later Android version. To get around this and not use something like PowerMock to mock static methods like Base64.encode(), I created the [Base64Wrapper](https://github.com/Azure/azure-sdk-for-android/blob/67845f0155a5700833dae754fbd7c710ca8aaddb/sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java#L94) interface. Normally it will always be set to [Base64Android](https://github.com/Azure/azure-sdk-for-android/blob/67845f0155a5700833dae754fbd7c710ca8aaddb/sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java#L101) and for unit tests it will be set to [Base64Stub](https://github.com/Azure/azure-sdk-for-android/blob/67845f0155a5700833dae754fbd7c710ca8aaddb/sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64Stub.java).

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java (hunk 1 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java]: Would it be better to move the interface and class below to their own files?

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java (hunk 6 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java]: You are right, this can be made to use `assertNull`.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java (hunk 7 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UrlTest.java]: Here we can just use `assertNotEquals`.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UtilTest.java (hunk 8 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UtilTest.java]: Sounds good, will add some tests that uses `assertNonNull`.

Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UtilTest.java]: Based on the test names, it actually seems I was planning to have the non-null tests but somehow forgot about it.

Comment (vcolin7) [reply sdk/core/azure-core/src/test/java/com/azure/android/core/util/Base64UtilTest.java]: I will correct and add tests that check for the positive scenarios when we have a non-null source and non-empty string.

**Comment by vcolin7** [comment]
> 
> 
> Are any of these tests portable to `azure-sdk-for-java` and cover gaps in testing there? If so would you mind opening either an issue or a PR against the repo and either copy or reference this PR.

I'm pretty sure some of these should be portable to the Java repo. I'll create a PR later during the week.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java]: Leaving them in this file is fine, I just didn't have a complete understanding of this when I wrote the comment.
